slf4j-internal/src/main/scala/io/chrisdavenport/log4cats/slf4j/internal/LoggerMacros.scala

```diff
@@ -0,0 +1,208 @@
+        param)}))($f)"
+
+    def loggerBySymbolName(s: Symbol)(f: c.Expr[F]) = {
+      def fullName(s: Symbol): String = {
+        @inline def isPackageObject = (
+          (s.isModule || s.isModuleClass)
+            && s.owner.isPackage
+            && s.name.decodedName.toString == termNames.PACKAGE.decodedName.toString
+        )
+        if (s.isModule || s.isClass) {
+          if (isPackageObject) {
+            s.owner.fullName
+          } else if (s.owner.isStatic) {
+            s.fullName
+          } else {
+            fullName(s.owner) + "." + s.name.encodedName.toString
+          }
+        } else {
+          fullName(s.owner)
+        }
+      }
+      loggerByParam(q"${fullName(s)}")(f)
+    }
+
+    def loggerByType(s: Symbol)(f: c.Expr[F]) = {
+      val typeSymbol: ClassSymbol = (if (s.isModule) s.asModule.moduleClass else s).asClass
+      val typeParams = typeSymbol.typeParams
+
+      if (typeParams.isEmpty) {
+        loggerByParam(q"classOf[$typeSymbol]")(f)
+      } else {
+        if (typeParams.exists(_.asType.typeParams.nonEmpty)) {
+          /* We have at least one higher-kinded type: fall back to by-name logger construction, as
+           * there's no simple way to declare a higher-kinded type with an "any" parameter. */
+          loggerBySymbolName(s)(f)
+        } else {
+          val typeArgs = List.fill(typeParams.length)(WildcardType)
+          val typeConstructor = tq"$typeSymbol[..${typeArgs}]"
+          loggerByParam(q"classOf[$typeConstructor]")(f)
+        }
+      }
+    }
+
+    @inline def isInnerClass(s: Symbol) =
+      s.isClass && !(s.owner.isPackage)
+
+    val instanceByName = Slf4jLoggerInternal.singletonsByName && (cls.isModule || cls.isModuleClass) || cls.isClass && isInnerClass(
+      cls
+    )
+
+    if (instanceByName) {
+      loggerBySymbolName(cls)(f)
+    } else {
+      loggerByType(cls)(f)
+    }
+  }
+
+  /** A macro context that represents a method call on a Logger instance. */
+  private[this] type LogCtx[F[_]] = whitebox.Context { type PrefixType = Slf4jLoggerInternal[F] }
+
+  /** Log a message reflectively at a given level.
+   *
+   * This is the internal workhorse method that does most of the logging for real applications.
+   *
+   * @param msg the message that the user wants to log
+   * @param error the `Throwable` that we're logging along with the message, if any
+   * @param logLevel the level of the logging
+   */
+  private[this] def reflectiveLog[F[_]](
+      c: LogCtx[F]
+  )(msg: c.Tree, error: Option[c.Expr[Throwable]], context: Seq[c.Expr[(String, String)]])(
+      logLevel: LogLevel) = {
+    import c.universe._
+
+    val logger = q"${c.prefix.tree}.logger"
+    val F = q"${c.prefix}.F"
+    val logValues = error match {
+      case None => List(msg)
+      case Some(e) => List(msg, e.tree)
+    }
+    val logExpr = q"$logger.${TermName(logLevel.methodName)}(..$logValues)"
+    val checkExpr = q"$logger.${TermName(s"is${logLevel.name}Enabled")}"
```

Comment:
This calls `.isEnabled` on the java logger. I didn't delay this in an `F[_]` because from the looks of it it seemed to be referentially transparent once the logger is created. Creation of the logger itself being the impure operation.

But I wasn't 100% sure 😊

Reply:
By what criteria is the creation of the logger not referentially transparent? As in because it loads resources. In that case do you think the methods should return `F[Logger[F]]`

Reply:
@ChristopherDavenport, yes I was thinking about the fact that logger creation loads resources, therefore it is unsafe. 

We could leave current methods as they are (since I have a hunch that even all us FP people still use unsafe creation methods), and maybe add additional methods of the form:`safe${factoryMethod} : F[Logger[F]]`.

This latter approach is slightly problematic since the convention is usually the other way around: adding the `unsafe` word to method names. Therefore, the second option would be to stick to convention, and basically ruin all current factory methods, and provide the same functionality under `unsafe${factoryMethod}: Logger[F]` method names.

I'm also fine with not doing anything else 🤷‍♂️. I have never had problems with the impurity of logger creation, only with the impurity of logging itself 😄.

Reply:
I agree. I think I'll throw these into an unsafe name, and with a corresponding equivalent one in `F` just so that we are totally transparent.